### ISSUE TITLE
Incorrect guard for intersection types

### ISSUE THREAD

**Issue by mariosaban** [rhys-vdw/ts-auto-guard#172]
Given the following type
```
export type MyType = {
    foo: string;
} & {
    bar: number;
};
```

when running ts-auto-guard (version 1.0.0) it creates the following guard function:
```
export function isMyType(obj: any, _argumentName?: string): obj is MyType {
    return (
        (((obj !== null && typeof obj === 'object') || typeof obj === 'function') && typeof obj.foo === 'string') ||
        (((obj !== null && typeof obj === 'object') || typeof obj === 'function') && typeof obj.bar === 'number')
    );
}
```

But I would expect the following guard function to be generated:
```
export function isMyType(obj: any, _argumentName?: string): obj is MyType {
    return (
        (((obj !== null && typeof obj === 'object') || typeof obj === 'function') && typeof obj.foo === 'string') &&
        (((obj !== null && typeof obj === 'object') || typeof obj === 'function') && typeof obj.bar === 'number')
    );
}
```
where the conditions for both subtypes are combined via && instead of || because the type constraints of all subtypes have to be matched. With the currently generated guard the code `isMyType({foo: 'hello'})` would return `true` although the required property `bar` is missing.

**Comment by rhys-vdw** [comment]
Yes, indeed. This can replace #28 because it's far more informative.

Wanna give it a shot?

**Comment by rhys-vdw** [comment]
Closed by #176
